### ISSUE TITLE
Question: no-object-literal-type-assertion and arrow functions

### ISSUE THREAD

**Issue by danvk** [palantir/tslint#2724]
I like the new `no-object-literal-type-assertion` rule, but it's a bit awkward in the context of arrow functions.

For example, consider this code:

```ts
interface MyType {
  constant: 'c';
}

function foo(blah: string): MyType[] {
  return blah.split('').map((char, i) => ({constant: 'c'}));
}
```

This fails to type check:
```
[ts]
Type '{ constant: string; }[]' is not assignable to type 'MyType[]'.
  Type '{ constant: string; }' is not assignable to type 'MyType'.
    Types of property 'constant' are incompatible.
      Type 'string' is not assignable to type '"c"'.
```

Adding a cast fixes it:

```ts
function foo(blah: string): MyType[] {
  return blah.split('').map((char, i) => ({constant: 'c'} as MyType));
}
```

But this `no-object-literal-type-assertion` requires me to rewrite it as:

```ts

function foo(blah: string): MyType[] {
  return blah.split('').map((char, i) => {
    const c: MyType = {constant: 'c'};
    return c;
  });
}
```

I realize that the latter form is [more correct][1]. But it's sad to lose the compact arrow return shorthand syntax. Is there a way to get both?

[1]: http://stackoverflow.com/questions/41964087/in-typescript-is-there-a-reason-to-prefer-const-foo-type-or-const-foo

**Comment by adidahiya** [comment]
Hmm, to me this seems like more of an issue around the lack of ergonomics around string literal types. If the interface uses an enum instead, it works:

```ts
enum Constant {
  A, B, C
}


interface MyType {
  constant: Constant;
}

function foo(blah: string): MyType[] {
  return blah.split('').map((char, i) => ({constant: Constant.C}));
}
```

If you add a bit of infrastructure like [typescript-string-enums](https://github.com/dphilipson/typescript-string-enums) then you can get the benefits of strings + enums.

Alternatively, you can also limit the places where you're casting:

```ts
function foo(blah: string): MyType[] {
  return blah.split('').map((char, i) => ({constant: 'c' as 'c'}));
}
```

... but that looks awkward. I'm not sure if we should relax the rule for this case, but am open to discussion & additional feedback around this.

**Comment by danvk** [comment]
In my code, the type in question is a [GeoJSON feature][1], so migrating to something like typescript-string-enums isn't really an option. The `'c' as 'c'` solution would work, but like you say it's pretty awkward.

I wouldn't suggest relaxing the rule in this case, either, since it's exactly the sort of issue you'd like to catch. Maybe we should file / thumbs up a TypeScript issue to deduce string literal types better in situations like this.

[1]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/46ad6f983339ca888845ef9eee8cffe0060b9556/types/geojson/index.d.ts#L93

**Comment by adidahiya** [comment]
yeah, so basically we could try to improve all the existing typings out there to move away from string literals (like types/geojson), or get tsc to be smarter

**Comment by JoshuaKGoldberg** [comment]
Sounds like this is out of scope for TSLint itself. Thanks for the discussion!
